### PR TITLE
[GOBBLIN-2157] Copy table properties in iceberg distcp

### PR DESCRIPTION
**Pull request by Will-Lo**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [ ] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-2157

### Description
- [ ] Here are some details about my PR, including screenshots (if applicable):
The existing iceberg registration step has the following snippet:
```
    if (dstMetadata != null) {
      // use current destination metadata as 'base metadata' and source as 'updated metadata' while committing
      this.tableOps.commit(dstMetadata, srcMetadata.replaceProperties(dstMetadata.properties())); <<<<<------- BUG
    }
```
Since Iceberg's `replaceProperties` https://github.com/apache/iceberg/blob/e449d3405cfdb304c94835845bd8f34a73b4a517/core/src/main/java/org/apache/iceberg/TableMetadata.java#L583 just replaces the full set of properties, not just any existing properties, the result is that all of the properties are equivalent to the destination table metadata properties. Which effectively copies no properties.

This PR modifies this behavior so that the destination table properties can be overwritten by the source table properties if the source table updates existing properties, as well as maintaining any existing properties in the destination that is not defined on the source.

### Tests
- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
Unit tests

### Commits
- [ ] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/iceberg/IcebergTable.java [hunk 1]
```diff
@@ -213,7 +215,10 @@ public DatasetDescriptor getDatasetDescriptor(FileSystem fs) {
   protected void registerIcebergTable(TableMetadata srcMetadata, TableMetadata dstMetadata) {
     if (dstMetadata != null) {
       // use current destination metadata as 'base metadata' and source as 'updated metadata' while committing
-      this.tableOps.commit(dstMetadata, srcMetadata.replaceProperties(dstMetadata.properties()));
+      Map<String, String> combinedMetadataProperties = Maps.newHashMap();
+      combinedMetadataProperties.putAll(dstMetadata.properties());
+      combinedMetadataProperties.putAll(srcMetadata.properties());
```
Comment (phet) [inline gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/iceberg/IcebergTable.java:220]: please add a comment that captures the reasoning behind overshadowing dest with src

**Comment by Blazer-007** [comment]
@Will-Lo I believe if we want to copy some source side table properties then we should copy only selective property not all,
 
As lets say we change some property like clusterId or tableId or any other and that property is copied over to destination but destination requires this doesn't to be changed then in this case we will eventually be corrupting the properties.

**Review by phet — APPROVED** [review]

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/iceberg/IcebergTable.java [hunk 2]
```diff
@@ -212,8 +212,9 @@ public DatasetDescriptor getDatasetDescriptor(FileSystem fs) {
    * @param dstMetadata is null if destination {@link IcebergTable} is absent, in which case registration is skipped */
   protected void registerIcebergTable(TableMetadata srcMetadata, TableMetadata dstMetadata) {
     if (dstMetadata != null) {
-      // use current destination metadata as 'base metadata' and source as 'updated metadata' while committing
-      this.tableOps.commit(dstMetadata, srcMetadata.replaceProperties(dstMetadata.properties()));
+      // Use current destination metadata as 'base metadata' and but update to source metadata when committing
+      // If any of the source table properties are deleted, they will be reflected in the destination table
```
Comment (phet) [inline gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/iceberg/IcebergTable.java:216]: "and but"

how about - 
> ...as 'base metadata', but commit the source-side metadata, to synchronize source-side property deletion over to the destination

?

File: gobblin-data-management/src/test/java/org/apache/gobblin/data/management/copy/iceberg/IcebergTableTest.java [hunk 3]
```diff
@@ -181,6 +184,33 @@ public void testGetIncrementalSnapshotInfosIteratorRepeatedFiles() throws IOExce
...
+  @Test
+  public void testNewTablePropertiesAreRegistered() throws Exception {
+    Map<String, String> srcTableProperties = Maps.newHashMap();
+    Map<String, String> destTableProperties = Maps.newHashMap();
+
+    srcTableProperties.put("newKey", "newValue");
+    // Expect the old value to be overwritten by the new value
+    srcTableProperties.put("testKey", "testValueNew");
+    destTableProperties.put("testKey", "testValueOld");
+    // Expect existing property values to be maintained if it does not exist on the source
```
Comment (phet) [inline gobblin-data-management/src/test/java/org/apache/gobblin/data/management/copy/iceberg/IcebergTableTest.java:197]: comment looks out-of-date

**Review by phet — APPROVED** [review]

**Pull request merged**